routing/router.go

```diff
@@ -1744,7 +1746,14 @@ func (r *ChannelRouter) sendPaymentAttempt(paySession *paymentSession,
 
 	switch result := result.(type) {
 	case *htlcswitch.PaymentSuccess:
-		return result.Preimage, true, nil
+		preimage := result.Preimage
+		err := r.savePayment(p.route, preimage[:])
```

Comment:
The router should save the payment data _before_ the attempt, then only update the final route and preimage here. This will also allow us to eventually cancel/resume payments as well.

Reply:
We should also add tests for this new persistence move.

Reply:
> The router should save the payment data before the attempt, then only update the final route and preimage here. 

Done: https://github.com/lightningnetwork/lnd/pull/2761/commits/148afd04bb3016855452004b90fd0c568f9d4bb1

> This will also allow us to eventually cancel/resume payments as well.

This require us to store a lot more information, and is intentionally left out of this PR.